sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/predicates.scala

```diff
@@ -107,6 +107,14 @@ trait PredicateHelper extends AliasHelper with Logging {
     }
   }
 
+  /**
+   * First split the predicates by And, then combine the predicates with the same references.
```

Comment:
I think there is a misunderstanding. What I proposed is `splitConjunctivePredicates` should split AND/OR under `With`: `With(c1, c2, c1 + c1 > 0 AND c2 + c2 > 0 AND c1 + c2 > 0)` can be split into `With(c1, c1 + c1 > 0)`, `With(c2, c2 + c2 > 0)`, ` With(c1, c2, c1 + c2 > 0)`. When combine the predicates later, we also recognize the `With` and merge them to avoid duplicated expressions.

Reply:
why split OR

Reply:
Merge `With` requires a certain cost. It seems that we only need to merge once before rewriting `With`.

Reply:
note: when splitting the predicates under `With` (by `AND`), we don't refresh the CTE ids, so when merging the `With` back, it should be pretty easy.

Reply:
It is unnecessary to continuously split and merge with. We only need to split by and to generate with, and finally merge with before rewriting with.